File: 89_Problem_2.cpp
// code_report 
// 

#include <vector>
#include <map>

using namespace std;

int carFleet (int target, vector<int>& p, vector<int>& s)
{
   map<float, float> cars;
   for (int i = 0; i < p.size (); ++i) cars.insert ({ -p[i], s[i] });

   auto fleets    = 0;
   auto last_time = -1.0f;

   for (const auto& car : cars) {
      auto time = (target + car.first) / car.second;
      if (time > last_time) fleets++, last_time = time;
   }

   return fleets;
}

File: 89_Problem_2.java
// code_report
// https://youtu.be/H5w6doOXz10

import java.util.*;
import java.io.*;

public int carFleet(int target, int[] p, int[] s) 
{
    TreeMap<Float,Float> cars = new TreeMap ();
    for (int i = 0; i < p.length; ++i) cars.put (-(float)p[i], (float)s[i]);
    
    Integer fleets    = 0;
    Float   last_time = 0.0f;
    
    for (Map.Entry<Float,Float> car : cars.entrySet ()) {
        Float time = (target + car.getKey ()) / car.getValue ();
        if (time > last_time) {
            fleets++;
            last_time = time;
        }
    }
    
    return fleets;
}
